### PR TITLE
[JIT] Fix classes as attributes in recursive scripting

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32594 [JIT] Fix classes as attributes in recursive scripting**

Differential Revision: [D19562951](https://our.internmc.facebook.com/intern/diff/D19562951)

### CODE REVIEW

File: torch/csrc/jit/pybind_utils.h [hunk 1]
```diff
@@ -153,6 +153,24 @@ inline InferredType tryToInferType(py::handle input) {
...
+    py::str qualifiedName = py::module::import("torch.jit")
+                                .attr("_qualified_name")(input.get_type());
+    auto pyClass = py::module::import("torch.jit")
+                       .attr("_get_script_class")(qualifiedName);
+    if (!pyClass.is_none()) {
+      auto cu = get_python_cu();
+      const auto classname =
+          c10::QualifiedName(py::cast<std::string>(qualifiedName));
+      auto class_type = cu->get_class(classname);
+      if (class_type) {
```
Comment (suo) [inline torch/csrc/jit/pybind_utils.h]: This should be a hard error I think. If a class is in the `_get_script_class` mapping but not in the Python CU, something is broken.

**Review by suo — APPROVED** [review]
what ever happened to ok boomer

File: torch/csrc/jit/pybind_utils.h (hunk 1 above)
Comment (jamesr66a) [reply torch/csrc/jit/pybind_utils.h]: sure

File: test/test_jit.py [hunk 2]
```diff
@@ -4981,6 +4981,23 @@ def foo():
...
+
+        class FooBar1234(torch.nn.Module):
+            def __init__(self):
+                super(FooBar1234, self).__init__()
+                self.f = Foo321()
+
+            def forward(self, x):
+                return x + self.f.x
+
+        scripted = torch.jit.script(FooBar1234())
```
Comment (eellison) [inline test/test_jit.py]: Could you try to run the model?

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 44977395:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=449773956bae5b89f1d9ab31b42cbe8f311d0c69).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4335435) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90479067) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90479067))
<details>
<summary>
<code>Jan 24 22:13:40 RuntimeError: test_jit_simple failed!</code>
</summary>

```
Jan 24 22:13:40 Ran 2360 tests in 84.293s 
Jan 24 22:13:40  
Jan 24 22:13:40 FAILED (errors=1, skipped=79, expected failures=1) 
Jan 24 22:13:40  
Jan 24 22:13:40 Generating XML reports... 
Jan 24 22:13:40 Traceback (most recent call last): 
Jan 24 22:13:40   File "test/run_test.py", line 456, in <module> 
Jan 24 22:13:40     main() 
Jan 24 22:13:40   File "test/run_test.py", line 449, in main 
Jan 24 22:13:40     raise RuntimeError(message) 
Jan 24 22:13:40 RuntimeError: test_jit_simple failed! 
Jan 24 22:13:40 + cleanup 
Jan 24 22:13:40 + retcode=1 
Jan 24 22:13:40 + set +x 
Jan 24 22:13:40 =================== sccache compilation log =================== 
Jan 24 22:13:40 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 24 22:13:40 Compile requests                 2 
Jan 24 22:13:40 Compile requests executed        0 
Jan 24 22:13:40 Cache hits                       0 
Jan 24 22:13:40 Cache misses                     0 
Jan 24 22:13:40 Cache timeouts                   0 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4335438) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_legacy_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90479258) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90479258))
<details>
<summary>
<code>Jan 24 22:14:24 RuntimeError: test_jit_legacy failed!</code>
</summary>

```
Jan 24 22:14:24 Ran 2360 tests in 131.614s 
Jan 24 22:14:24  
Jan 24 22:14:24 FAILED (errors=1, skipped=62, expected failures=1) 
Jan 24 22:14:24  
Jan 24 22:14:24 Generating XML reports... 
Jan 24 22:14:24 Traceback (most recent call last): 
Jan 24 22:14:24   File "test/run_test.py", line 456, in <module> 
Jan 24 22:14:24     main() 
Jan 24 22:14:24   File "test/run_test.py", line 449, in main 
Jan 24 22:14:24     raise RuntimeError(message) 
Jan 24 22:14:24 RuntimeError: test_jit_legacy failed! 
Jan 24 22:14:24 + cleanup 
Jan 24 22:14:24 + retcode=1 
Jan 24 22:14:24 + set +x 
Jan 24 22:14:24 =================== sccache compilation log =================== 
Jan 24 22:14:24 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 24 22:14:24 Compile requests               260 
Jan 24 22:14:24 Compile requests executed        0 
Jan 24 22:14:24 Cache hits                       0 
Jan 24 22:14:24 Cache misses                     0 
Jan 24 22:14:24 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/csrc/jit/pybind_utils.h [hunk 3]
```diff
@@ -693,6 +710,13 @@ inline py::object toPyObject(IValue ivalue) {
     AT_ASSERT(classType);
     auto pyClass =
         py::module::import("torch.jit").attr("_get_script_class")(obj->name());
```
Comment (dzhulgakov) [inline torch/csrc/jit/pybind_utils.h]: given tat it's in the hot path - shall we import torch.jit and lookup the method only once somewhere above? (not sure whether it matters)

File: torch/csrc/jit/pybind_utils.h (hunk 1 above)
Comment (jamesr66a) [reply torch/csrc/jit/pybind_utils.h]: This thing fails without it

**Comment by jamesr66a** [comment]
@suo I'm going to separate this out from my stack since 1) it actually isn't required for torchbind classes and 2) this weird error is blocking it

**Pull request closed without merging**